xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/exceptions/ExceptionExceptionMapper.java

```diff
@@ -0,0 +1,117 @@
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.rest.internal.exceptions;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.script.ScriptContext;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.localization.ContextualLocalizationManager;
+import org.xwiki.rest.XWikiRestComponent;
+import org.xwiki.script.ScriptContextManager;
+import org.xwiki.template.TemplateManager;
+
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN_TYPE;
+import static org.apache.commons.lang3.exception.ExceptionUtils.getRootCauseMessage;
+import static org.apache.commons.lang3.exception.ExceptionUtils.getStackTrace;
+
+/**
+ * Exception mapper for all {@link Exception} exceptions. Allowing to log all exceptions that are not inheriting from
+ * {@link RuntimeException} or already mapped by another exception mapper.
+ * <p>
+ * The goal is to be able to have access to the stacktrace of the exceptions, instead of a simple http 500 error and a
+ * {@code "No ExceptionMapper was found, but must be found"} message (see
+ * <a href="https://github.com/restlet/restlet-framework-java/issues/1070">
+ * https://github.com/restlet/restlet-framework-java/issues/1070</a>).
+ *
+ * @version $Id$
+ * @since 13.8RC1
+ * @since 13.7.1
+ * @since 13.4.4
+ */
+@Component
+@Named("org.xwiki.rest.internal.exceptions.ExceptionExceptionMapper")
+@Provider
+@Singleton
+public class ExceptionExceptionMapper implements ExceptionMapper<Exception>, XWikiRestComponent
+{
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private TemplateManager templateManager;
+
+    @Inject
+    private ScriptContextManager scriptContextManager;
+
+    @Inject
+    private ContextualLocalizationManager contextLocalization;
+
+    @Override
+    public Response toResponse(Exception cause)
+    {
+        // This ExceptionMapper captures too many exception types, because we can't distinguish checked and unchecked
+        // exceptions using generic types. If the mapped type is RuntimeException (unchecked exception), JaxRsProviders
+        // would re-throw the exception without mapping it (or it would have been handled by another mapper instead of
+        // this one), we reproduce this behavior to stay as close as possible to the default implementation (note:
+        // RuntimeException stacktraces are logged by default).
+        if (cause instanceof RuntimeException) {
+            throw (RuntimeException) cause;
+        }
+
+        // Logs the unmapped exception, with its stacktrace.
+        this.logger.error("A REST endpoint failed with an unmapped exception.", cause);
```

Comment:
Do we agree that if this code is reached it's that something unexpected happened and we can't recover from it? I'm asking to decide if `error` is good or if `warning` should be used. This means that something needs to be fixed and it's not a usage problem.

Reply:
Ending up here definitely means unexpected exception for me. Expected/document REST API behavior (so, something an HTTP client can use) is generally supposed to produce standard HTTP code (401, 404, etc.).

Reply:
Yes, exceptions thrown by REST endpoints needs to have an exception mapper, either by inheriting an exception with an exception mapper, or by defining a new one. This logs is only displayed when the exception is not handled properly by a REST endpoint.